src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,184 @@
+                this.key.Metadata = metadata;
+                if (string.IsNullOrEmpty(this.key.ETag))
+                    batchOperation.Insert(this.key);
+                else
+                    batchOperation.Replace(this.key);
+
+                // add new states
+                List<Tuple<string,long>> stored = this.states.Select(s => Tuple.Create(s.TransactionId, s.SequenceId)).ToList();
+                List<StateEntity> newStates = new List<StateEntity>();
+                foreach (PendingTransactionState<TState> pendingState in statesToPrepare.Where(p => !stored.Contains(Tuple.Create(p.TransactionId, p.SequenceId))))
+                {
+                    var newState = StateEntity.Create(this.JsonSettings, this.partition, pendingState);
+                    newStates.Add(newState);
+                    batchOperation.Insert(newState);
+                }
+
+                if (batchOperation.Count > AzureTableConstants.MaxBatchSize)
+                {
+                    this.logger.LogError("Too many pending states. PendingStateCount {PendingStateCount}.", batchOperation.Count);
+                    throw new InvalidOperationException($"Too many pending states. PendingStateCount {batchOperation.Count}");
+                }
+
+                await table.ExecuteBatchAsync(batchOperation).ConfigureAwait(false);
+                this.states.AddRange(newStates);
+                return this.key.ETag;
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError("Transactional state persist failed {Exception}.", ex);
+                throw;
+            }
+        }
+
+        public async Task<string> Confirm(string expectedETag, string metadata, string transactionIdToCommit)
+        {
+            try
+            {
+                // only update storage if transaction id is greater then previously committed
+                if (string.Compare(transactionIdToCommit, this.key.CommittedTransactionId) <= 0)
+                    return this.key.ETag;
+
+                if (!TryFindState(transactionIdToCommit, out TState state))
+                {
+                    this.logger.LogCritical("Transactional state non-recoverable error.  Attempting to confirm a transaction {TransactionId} for which no state exists.", transactionIdToCommit);
+                    throw new InvalidOperationException($"Transactional state non-recoverable error.  Attempting to confirm a transaction {transactionIdToCommit} for which no state exists.");
+                }
+
+                this.key.ETag = expectedETag;
+                this.key.Metadata = metadata;
+                this.key.CommittedTransactionId = transactionIdToCommit;
+                await WriteKey();
+                var dead = this.states.Where(p => string.Compare(p.TransactionId, transactionIdToCommit) <0).ToList();
+                this.states = this.states.Where(p => string.Compare(p.TransactionId, transactionIdToCommit) >= 0).ToList();
```

Comment:
We consider these APIs 'pubternal', I think - i.e, they're open but not really intended for users to implement until we're stable. Otherwise I wonder if it would be worth lifting some logic (like this cleanup) to a different layer.

Reply:
The ITransactionalStateStorage and ITransactionalStateStorageFactory interfaces are public, and we do expect custom implementations.  The patterns are still in flux, but the entities and cleanup logic here is specific to this implemetation.  Do you disagree?

Reply:
The cleanup logic may be semi-specific, but I think most providers would need/want some kind of cleanup process.

Anyhow, users will know to copy this implementation.

Reply:
I agree with your suspicions, but don't think we've enough information to draw such conclusions at this time.  Some systems may not cleanup at all, but instead perform more targeted loads, others may archive the states for transaction history.  Some may keep all state in a single object (blob storage).  Even if cleanup is desired, the pattern I'm using is debatable. 

I advocate keeping the public surface as simple and direct as is practicable and geared towards the needs of the transaction orchestration.

Reply:
Seems ok to me for now.